traits/observers/_filtered_trait_observer.py

```diff
@@ -0,0 +1,186 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+from traits.observers._has_traits_helpers import (
+    iter_objects,
+    observer_change_handler,
+)
+from traits.observers._i_observer import IObserver
+from traits.observers._observer_change_notifier import ObserverChangeNotifier
+from traits.observers._trait_change_event import trait_event_factory
+from traits.observers._trait_event_notifier import TraitEventNotifier
+from traits.trait_base import Uninitialized
+
+
+@IObserver.register
+class FilteredTraitObserver:
+    """ An observer for observing traits using a custom filter.
+
+    Parameters
+    ----------
+    notify : boolean
+        Whether to notify for changes.
+    filter : callable(str, CTrait) -> boolean
+        A callable that receives the name of a trait and the corresponding
+        trait definition. The returned boolean indicates whether the trait is
+        observed. In order to remove an existing observer with the equivalent
+        filter, the filter callables must compare equally.
```

Comment:
Perhaps also worth documenting that the filter has to be hashable? (Or maybe not, since it's going to be true in practice for functions and methods; you'd probably have to try quite hard to create a non-hashable callable.)